### PR TITLE
PDT-618 Weighted replica round robin

### PR DESCRIPTION
**Pull request by nicwolff**
We don't want to read so much from the Redis primary, and when AWS ElastiCache promotes a Redis replica to primary, we want to read less from it and start reading more from its replacement.

(We still want to read _some_ from the primary, to trigger eviction of expired keys that will get replicated to the replicas – Redis 2.8 doesn't expire on replicas ಠ_ಠ)

This change attempts to identify the primary by comparing the IP address we get by resolving its hostname alias to the IPs of the replicas. On failover, we assume that the DNS cache is clear and we can find the new primary the same way immediately.

Then it stocks the list of Redis objects to read from with multiple pointers to each replica, so that more reads will go to the replicas.

This is hard to test in Docker since even if you set up a Redis cluster you don't have the DNS alias switching. We plan to deploy it to one of the EC2 environments for testing.

### CODE REVIEW

**Comment by oharlem** [comment]
@nicwolff 
"...On failover, we assume that the DNS cache is clear..."
Do we flush DNS cache on failover? 
Why do we assume that?
What happens if DNS cache is stale?

File: cacheops/redis.py
Comment (oharlem) [inline cacheops/redis.py]: In what cases this is a string?
I see a list returned as per `https://github.com/HearstCorp/rover/blob/develop/rover/settings/ec2.py#L26`

File: cacheops/redis.py
Comment (oharlem) [inline cacheops/redis.py]: @nicwolff 
I do not see any additional info on settings.REDIS_REPLICA_WEIGHT.
Is this something we forgot to initialize 
or coming from some additional work that needs to be done (need to mention this in the PR descr + "Do Not Merge" label + comment to the ticket if dependency on that)
or smth else?

Comment (oharlem) [inline cacheops/redis.py]: Why not add a required # of replica "copies" directly to `settings.REDIS_REPLICAS`?
This way:
1. We'll be able to assign different weights to different slaves if necessary
2. Will avoid this extra iteration
3. Anyways, number of duplicates is not something computed, but set through the config.
Thoughts?

File: cacheops/redis.py
Comment (oharlem) [inline cacheops/redis.py]: `from_url` returns a client object as per `https://redis-py.readthedocs.io/en/latest/#redis.StrictRedis.from_url`. Is this what we want here?
Another concern: cacheops uses db=1, while from_url will give db=0 by default.

**Review by oharlem — CHANGES_REQUESTED** [review]
@nicwolff 
No changes required (misleading GitHub's review type), but let's touch base regarding my comments first before merging. Thanks!

File: cacheops/redis.py
Comment (nicwolff) [reply cacheops/redis.py]: The settings come from Rover.

File: cacheops/redis.py
Comment (nicwolff) [reply cacheops/redis.py]: This is in the existing code, there's no reason to change it here.

File: cacheops/redis.py
Comment (nicwolff) [reply cacheops/redis.py]: The whole point of this is that the primary's `CNAME` will change IPs on failover – one of the hostnames in the `REDIS_REPLICAS` list is actually the primary, but we don't know which until this code runs. So 3 is false, the number of duplicates of each hostname _is_ computed, and we can't do this in the config.

File: cacheops/redis.py
Comment (nicwolff) [reply cacheops/redis.py]: Yes, `redis_replicas` is a list of Redis client objects – see https://github.com/HearstCorp/django-cacheops/blob/PDT-618-weighted-replica-round-robin/cacheops/redis.py#L107 where we use one to `.get()` from the cache:
```py
    class ReplicaProxyRedis(CacheopsRedis):
        """Proxy `get` calls to redis replica."""
        def get(self, *args, **kwargs):
            try:
                redis_replica = random.choice(redis_replicas)
                return redis_replica.get(*args, **kwargs)
            except redis.ConnectionError:
                return super(ReplicaProxyRedis, self).get(*args, **kwargs)
```

Comment (nicwolff) [reply cacheops/redis.py]: But you know what I am doing that's dumb – I'm adding the extra URLs then making client objects for all of them, instead of making client objects first and duplicating the references! Will add that...

**Comment by nicwolff** [comment]
The DNS cache question is a good one – it's how the current failover works, so in theory calling `socket.gethostbyname` will get us the new IP for the primary's `CNAME`. But I'm not sure, that's why this needs testing in EC2! I'm working with Nick on that now.

**Review by oharlem — APPROVED** [review]
Looks great.
Let's just keep in mind DNS when testing.

**Comment by nicwolff** [comment]
Indeed, failover doesn't seem to be working in EC2. I'm adding more logging and looking into that, but we have a current situation with replica evictions that the basic read-balancing that is working here could help, so I plan to deploy this and start another ticket for the failover.

(Redis replica promotions seem to be pretty rare anyway – we haven't logged one in production in the last 30 days.)

File: cacheops/redis.py
Comment (akarpodinis) [inline cacheops/redis.py]: These three lines are insanity distilled in code.  Can I trouble you to please add some comments to talk about what's happening here?

File: cacheops/redis.py
Comment (akarpodinis) [inline cacheops/redis.py]: Maybe `rebalance_replicas`?  I think that's what's going on here, and that's the word you used when you walked me through the logic a couple of minutes ago.

**Comment by nicwolff** [comment]
Made some readability changes suggested by @akarpodinis.

**Comment by nicwolff** [comment]
@akarpodinis @nmohoric CRs please!

**Review by akarpodinis — APPROVED** [review]

**Pull request merged**
